pkgs/build-support/make-desktopitem/default.nix

```diff
@@ -1,119 +1,207 @@
+        throw "makeDesktopItem: values in ${key} list must not contain semicolons!"
+      else if value == [ ] then
+        null
+      else
+        builtins.concatStringsSep ";" value;
+
+    # The [Desktop Entry] section of the desktop file, as an attribute set.
+    # Please keep in spec order.
+    mainSection = {
+      "Type" = type;
+      "Version" = "1.4";
+      "Name" = desktopName;
+      "GenericName" = genericName;
+      "NoDisplay" = boolOrNullToString noDisplay;
+      "Comment" = comment;
+      "Icon" = icon;
+      "OnlyShowIn" = renderList "onlyShowIn" onlyShowIn;
+      "NotShowIn" = renderList "notShowIn" notShowIn;
+      "DBusActivatable" = boolOrNullToString dbusActivatable;
+      "TryExec" = tryExec;
+      "Exec" = exec;
+      "Path" = path;
+      "Terminal" = boolOrNullToString terminal;
+      "Actions" = renderList "actions" (builtins.attrNames actions);
+      "MimeType" = renderList "mimeTypes" mimeTypes;
+      "Categories" = renderList "categories" categories;
+      "Implements" = renderList "implements" implements;
+      "Keywords" = renderList "keywords" keywords;
+      "StartupNotify" = boolOrNullToString startupNotify;
+      "StartupWMClass" = startupWMClass;
+      "URL" = url;
+      "PrefersNonDefaultGPU" = boolOrNullToString prefersNonDefaultGPU;
+      # "SingleMainWindow" = boolOrNullToString singleMainWindow;
+    } // extraConfig;
+
+    # Render a single attribute pair to a Key=Value line.
+    # FIXME: this isn't entirely correct for arbitrary strings, as some characters
+    # need to be escaped. There are currently none in nixpkgs though, so this is OK.
+    renderLine = name: value: if value != null then "${name}=${value}" else null;
+
+    # Render a full section of the file from an attrset.
+    # Null values are intentionally left out.
+    renderSection =
+      sectionName: attrs:
+      lib.pipe attrs [
+        (lib.mapAttrsToList renderLine)
+        (builtins.filter (v: v != null))
+        (builtins.concatStringsSep "\n")
+        (section: ''
+          [${sectionName}]
+          ${section}
+        '')
+      ];
+
+    mainSectionRendered = renderSection "Desktop Entry" mainSection;
+
+    # Convert from javaCase names as used in Nix to PascalCase as used in the spec.
+    preprocessAction =
+      {
+        name,
+        icon ? null,
+        exec ? null,
+      }:
+      {
+        "Name" = name;
+        "Icon" = icon;
+        "Exec" = exec;
+      };
+    renderAction = name: attrs: renderSection "Desktop Action ${name}" (preprocessAction attrs);
+    actionsRendered = lib.mapAttrsToList renderAction actions;
+
+    extension = if type == "Directory" then "directory" else "desktop";
+    content = [ mainSectionRendered ] ++ actionsRendered;
+  in
+  writeTextFile {
+    name = "${name}.${extension}";
+    destination = if lib.hasSuffix "/" destination then "${destination}${name}.${extension}" else "${destination}/${name}.${extension}";
```

Comment:
Would it be a better idea to throw instead if it has the suffix ? In that case, an assert statement can be added at the start of the function and this condition can be simplified.

Reply:
(Is even that necessary? Bash and therefore `writeTextFile` ignore doubled-up slashes.)

Reply:
> I'm not aware of that, and that's good news. We can remove that check here then !